rosbag2_cpp/src/rosbag2_cpp/cache/cache_consumer.cpp

```diff
@@ -53,11 +55,8 @@ void CacheConsumer::close()
 void CacheConsumer::change_consume_callback(
   CacheConsumer::consume_callback_function_t consume_callback)
 {
+  std::lock_guard<std::mutex> lg(consumer_mutex_);
   consume_callback_ = consume_callback;
-  if (!consumer_thread_.joinable()) {
```

Comment:
This is a required section. The cache consumer is closed when splitting bags - which means that we await the consumer thread to finish its work (flush the cache) and join it. We also enter the flushing state which is controlled by the flag.

If we don't restart the thread for the new bag/callback, we have no thread to do the work. Predictably, without this section the PR benchmarks results in all the messages being lost in subsequent bags (apart from bag 0).

This has to stay and it is a necessary condition. What could be improved is certainly some clarification around it.

Reply:
Oh, I see.  Yeah, that makes sense.

However, the previous code was buggy as well in that if you called `change_consumer_callback` without a `close`, you would "leak" the thread.

So I still think this class needs to be refactored.  One way to do it would be to make it totally RAII, where the constructor creates the thread, the destructor deletes the thread, and nothing else ever kills off the thread.  In that case, all `close` would do is to do a `message_cache->finalize()`, followed by `message_cache->notify_flushing_done`.  What do you think about that?

Reply:
Calling message_cache->finalize() and then message_cache->notify_flushing_done would be a logic error - flushing is not done until the consumer thread exits the loop. Closing a bag (whether it is the final one or not) should ensure the flush.

Reply:
I'm open to ideas.  At the end of the day, I'm not a huge fan of APIs that are easy to misuse, and this one seems to fit that description.  Do you have any thoughts on how we could make this nicer?

Reply:
With this PR merged, I don't think it is will be as easy to misuse anymore. Which part do you think is still easy to misuse?
Refactoring this, I would like to make sure #640 is solved - there is no reason not to use the cache while the consumer is finalizing, but some additional logic is needed.

Reply:
> Refactoring this, I would like to make sure #640 is solved - there is no reason not to use the cache while the consumer is finalizing, but some additional logic is needed.

All right.  I'm mostly interested in fixing the occasional crash we see in CI, so instead of getting bogged down with #640 I think we can instead fix the test to properly call `close`.  I'm going to change this PR to do that for now instead.

I'll leave it to you whether you want to refactor this as part of #640.  Thanks for the input so far.